libptable/ptable.c

```diff
@@ -0,0 +1,192 @@
+ * Phoenix-RTOS
+ *
+ * Partition table
+ *
+ * Copyright 2020, 2023 Phoenix Systems
+ * Author: Hubert Buczynski, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <ctype.h>
+#include <endian.h>
+#include <string.h>
+
+/* Per project board configuration may overwrite PTABLE_CHECKSUM definition */
+#include <board_config.h>
+
+#include "ptable.h"
+
+
+/* Enable checksum for partition table */
+#ifndef PTABLE_CHECKSUM
+#define PTABLE_CHECKSUM 0
+#endif
+
+
+static uint32_t ptable_crc32(uint32_t crc, const void *data, size_t len)
+{
+	int i;
+
+	while (len--) {
+		crc ^= *(char *)data++;
+		for (i = 0; i < 8; i++) {
+			crc = (crc >> 1) ^ ((crc & 1) ? 0xedb88320 : 0);
+		}
+	}
+
+	return crc;
+}
+
+
+static int ptable_partVerify(const ptable_t *ptable, const ptable_part_t *part, uint32_t memsz, uint32_t blksz)
+{
+	const ptable_part_t *p;
+	int i;
+
+#if PTABLE_CHECKSUM
+	/* Verify partition checksum */
+	if (part->crc != ptable_crc32(0, part, sizeof(ptable_part_t) - sizeof(uint32_t))) {
+		return -1;
+	}
+#endif
+
+	/* Verify offset and size */
+	if (part->size == 0) {
+		return -1;
+	}
+
+	if (part->size % blksz) {
+		return -1;
+	}
+
+	if (part->offset % blksz) {
+		return -1;
+	}
+
+	if (part->offset + part->size > memsz) {
+		return -1;
+	}
+
+	/* Verify partition type */
+	switch (part->type) {
+		case ptable_raw:
+		case ptable_jffs2:
+		case ptable_meterfs:
+			break;
+
+		default:
+			return -1;
+	}
+
+	/* Verify partition name */
+	for (i = 0; i < sizeof(part->name); i++) {
+		if (!isalnum(part->name[i])) {
+			break;
+		}
+	}
+
+	if ((i == 0) || (i >= sizeof(part->name)) || (part->name[i] != '\0')) {
+		return -1;
+	}
+
+	/* Compare against previous partitions */
+	for (p = ptable->parts; p != part; p++) {
+		/* Check for range overlap */
+		if ((part->offset <= p->offset + p->size - 1) && (part->offset + part->size - 1 >= p->offset)) {
+			return -1;
+		}
+
+		/* Check for name duplicate */
+		if (strcmp((const char *)part->name, (const char *)p->name) == 0) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+static int ptable_verify(const ptable_t *ptable, uint32_t memsz, uint32_t blksz)
+{
+	uint32_t i;
+
+#if PTABLE_CHECKSUM
+	/* Verify header checksum */
+	if (ptable->crc != ptable_crc32(0, ptable, sizeof(ptable_t) - sizeof(uint32_t))) {
+		return -1;
+	}
+#endif
+
+	/* Verify number of partitions */
+	if (ptable->count * sizeof(ptable_part_t) > blksz - sizeof(ptable_t) - sizeof(ptable_magic)) {
+		return -1;
+	}
+
+	/* Verify magic bytes */
+	if (memcmp((uint8_t *)ptable + ptable_size(ptable->count) - sizeof(ptable_magic), ptable_magic, sizeof(ptable_magic)) != 0) {
+		return -1;
+	}
+
+	/* Verify partitions */
+	for (i = 0; i < ptable->count; i++) {
+		if (ptable_partVerify(ptable, ptable->parts + i, memsz, blksz) < 0) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+int ptable_deserialize(ptable_t *ptable, uint32_t memsz, uint32_t blksz)
+{
+#if __BYTE_ORDER != __LITTLE_ENDIAN
+	uint32_t i;
```

Comment:
Because it is a library I suggest you add a condition:
```suggestion
int ptable_deserialize(ptable_t *ptable, uint32_t memsz, uint32_t blksz)
{
	if(ptable == NULL) {
		return -1;
	}

#if __BYTE_ORDER != __LITTLE_ENDIAN
	uint32_t i;
```
Same in `ptable_serialize()`.

Reply:
Good idea, added in the next force-push.